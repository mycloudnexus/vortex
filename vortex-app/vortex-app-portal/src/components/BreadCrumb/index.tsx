import type { ReactElement, ReactNode } from 'react'
import type { BreadcrumbItemType, BreadcrumbSeparatorType } from 'antd/es/breadcrumb/Breadcrumb'

import { Link, matchPath, useLocation } from 'react-router-dom'
import { IRouteObject, routes } from '@/routers'

import { Breadcrumb as AntBreadcrumb } from 'antd'
import { styled } from 'styled-components'
import { useAppStore } from '@/stores/app.store'

interface StyledLinkProps {
  isLast: boolean
  mainColor: string
}

const StyledLink = styled(Link).withConfig({
  shouldForwardProp: (prop) => !['isLast', 'mainColor'].includes(prop)
})<StyledLinkProps>`
  color: ${({ isLast, mainColor }) => (isLast ? '#000' : mainColor)} !important;
  font-weight: normal;
  &:hover {
    color: ${({ isLast, mainColor }) => (isLast ? mainColor : '#000')} !important;
  }
`

const StyledBreadCrumb = styled(AntBreadcrumb)`
  padding: 15px 10px;
`

const renderBreadCrumbItems = (
  route: Partial<BreadcrumbItemType & BreadcrumbSeparatorType>,
  _params: object,
  routes: Partial<BreadcrumbItemType & BreadcrumbSeparatorType>[],
  _paths: string[],
  mainColor: string
): ReactNode => {
  const last = routes.indexOf(route) === routes.length - 1
  return (
    <StyledLink to={route.path ?? ''} isLast={last} mainColor={mainColor}>
      {route.breadcrumbName}
    </StyledLink>
  )
}

const findMatchingRoute = (path: string, routes: IRouteObject[]): IRouteObject | null => {
  for (const route of routes) {
    if (route.path && matchPath({ path: route.path, end: true }, path)) {
      return route
    }

    if (route.children) {
      const matchingChild = findMatchingRoute(path, route.children)
      if (matchingChild) {
        return matchingChild
      }
    }
  }
  return null
}

const getBreadCrumbName = (path: string, routes: IRouteObject[]): string => {
  const route = findMatchingRoute(path, routes)
  if (!route) return ''
  if (route.breadCrumbName) return route.breadCrumbName
  const match = matchPath({ path: route.path!, end: true }, path)
  const dynamicSegment = match && Object.values(match.params)[0]
  return dynamicSegment ? `${dynamicSegment}` : path
}

const BreadCrumb = (): ReactElement => {
  const location = useLocation()
  const { mainColor } = useAppStore()
  const pathSegments = location.pathname === '/' ? [] : location.pathname.split('/').filter(Boolean)

  const breadCrumbsItems: Partial<BreadcrumbItemType & BreadcrumbSeparatorType>[] =
    pathSegments.length === 0
      ? [{ breadcrumbName: getBreadCrumbName('/', routes), path: '/' }]
      : pathSegments.map((_, index) => {
          const url = `/${pathSegments.slice(0, index + 1).join('/')}`
          return {
            breadcrumbName: decodeURIComponent(getBreadCrumbName(url, routes)),
            path: url
          }
        })

  if (breadCrumbsItems.length === 0) return <></>

  return (
    <StyledBreadCrumb
      separator='>'
      items={breadCrumbsItems}
      style={{ fontSize: '18px' }}
      itemRender={(route, params, routes, paths) => renderBreadCrumbItems(route, params, routes, paths, mainColor)}
    />
  )
}

export default BreadCrumb
