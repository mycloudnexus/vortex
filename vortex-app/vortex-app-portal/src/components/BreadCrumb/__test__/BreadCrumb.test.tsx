import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import BreadCrumb from '..'

jest.mock('@/stores/app.store', () => ({
  useAppStore: () => ({
    mainColor: 'blue'
  })
}))

jest.mock('@/routers', () => ({
  routes: [
    {
      path: '/home',
      breadCrumbName: 'Home',
      children: [
        {
          path: '/home/about',
          breadCrumbName: 'About'
        },
        {
          path: '/home/profile',
          breadCrumbName: 'Profile',
          children: [
            {
              path: '/home/profile/details',
              breadCrumbName: 'Details'
            }
          ]
        }
      ]
    },
    {
      path: '/contact',
      breadCrumbName: ''
    },
    {
      path: '/',
      breadCrumbName: 'Root'
    }
  ]
}))

describe('BreadCrumb Component', () => {
  it('should render the breadcrumb correctly for a given path', () => {
    const { getByText, baseElement } = render(
      <MemoryRouter initialEntries={['/home/about']}>
        <BreadCrumb />
      </MemoryRouter>
    )
    expect(getByText('Home')).toBeInTheDocument()
    expect(getByText('About')).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render breadcrumb for root path', () => {
    const { baseElement, getByText } = render(
      <MemoryRouter initialEntries={['/']}>
        <BreadCrumb />
      </MemoryRouter>
    )
    expect(getByText('Root')).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render multiple breadcrumb levels', () => {
    const { getByText, baseElement } = render(
      <MemoryRouter initialEntries={['/home/profile/details']}>
        <BreadCrumb />
      </MemoryRouter>
    )
    expect(getByText(/Home/i)).toBeInTheDocument()
    expect(getByText(/Profile/i)).toBeInTheDocument()
    expect(getByText(/Details/i)).toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('BreadCrumb Component should handle missing breadcrumb name gracefully', () => {
    const { queryByText, baseElement } = render(
      <MemoryRouter initialEntries={['/contact']}>
        <BreadCrumb />
      </MemoryRouter>
    )

    expect(queryByText('Contact')).not.toBeInTheDocument()
    expect(baseElement).toMatchSnapshot()
  })

  it('should render nothing for a path with no matching routes', () => {
    const { queryByText, baseElement } = render(
      <MemoryRouter initialEntries={['/nonexistent']}>
        <BreadCrumb />
      </MemoryRouter>
    )
    expect(queryByText('Home')).toBeNull()
    expect(queryByText('About')).toBeNull()
    expect(baseElement).toMatchSnapshot()
  })
})
