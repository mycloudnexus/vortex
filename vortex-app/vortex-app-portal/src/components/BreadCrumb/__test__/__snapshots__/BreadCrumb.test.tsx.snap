// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadCrumb Component BreadCrumb Component should handle missing breadcrumb name gracefully 1`] = `
<body>
  <div>
    <nav
      class="ant-breadcrumb sc-gtLWhw jESsVZ css-dev-only-do-not-override-ni1kz0"
      style="font-size: 18px;"
    >
      <ol>
        <li>
          <a
            class="sc-blHHSb emlGen"
            href="/contact"
          >
            /contact
          </a>
        </li>
      </ol>
    </nav>
  </div>
</body>
`;

exports[`BreadCrumb Component should render breadcrumb for root path 1`] = `
<body>
  <div>
    <nav
      class="ant-breadcrumb sc-gtLWhw jESsVZ css-dev-only-do-not-override-ni1kz0"
      style="font-size: 18px;"
    >
      <ol>
        <li>
          <a
            class="sc-blHHSb emlGen"
            href="/"
          >
            Root
          </a>
        </li>
      </ol>
    </nav>
  </div>
</body>
`;

exports[`BreadCrumb Component should render multiple breadcrumb levels 1`] = `
<body>
  <div>
    <nav
      class="ant-breadcrumb sc-gtLWhw jESsVZ css-dev-only-do-not-override-ni1kz0"
      style="font-size: 18px;"
    >
      <ol>
        <li>
          <a
            class="sc-blHHSb cVxirP"
            href="/home"
          >
            Home
          </a>
        </li>
        <li
          aria-hidden="true"
          class="ant-breadcrumb-separator"
        >
          &gt;
        </li>
        <li>
          <a
            class="sc-blHHSb cVxirP"
            href="/home/profile"
          >
            Profile
          </a>
        </li>
        <li
          aria-hidden="true"
          class="ant-breadcrumb-separator"
        >
          &gt;
        </li>
        <li>
          <a
            class="sc-blHHSb emlGen"
            href="/home/profile/details"
          >
            Details
          </a>
        </li>
      </ol>
    </nav>
  </div>
</body>
`;

exports[`BreadCrumb Component should render nothing for a path with no matching routes 1`] = `
<body>
  <div>
    <nav
      class="ant-breadcrumb sc-gtLWhw jESsVZ css-dev-only-do-not-override-ni1kz0"
      style="font-size: 18px;"
    >
      <ol>
        <li>
          <a
            class="sc-blHHSb emlGen"
            href="/nonexistent"
          />
        </li>
      </ol>
    </nav>
  </div>
</body>
`;

exports[`BreadCrumb Component should render the breadcrumb correctly for a given path 1`] = `
<body>
  <div>
    <nav
      class="ant-breadcrumb sc-gtLWhw jESsVZ css-dev-only-do-not-override-ni1kz0"
      style="font-size: 18px;"
    >
      <ol>
        <li>
          <a
            class="sc-blHHSb cVxirP"
            href="/home"
          >
            Home
          </a>
        </li>
        <li
          aria-hidden="true"
          class="ant-breadcrumb-separator"
        >
          &gt;
        </li>
        <li>
          <a
            class="sc-blHHSb emlGen"
            href="/home/about"
          >
            About
          </a>
        </li>
      </ol>
    </nav>
  </div>
</body>
`;
